File: addons/motion/zone.go
package motion

import (
	"image"
	"math"
	"nvr/pkg/ffmpeg"
)

type zones []*zone

func (z zones) analyze(frame1, frame2, diff []uint8, onActive func(int, float64)) {
	diffFrames(frame1, frame2, diff)
	for i, zone := range z {
		if zone == nil {
			continue
		}
		score, isActive := zone.checkDiff(diff)
		// score, active := zone.compareFrames(frame1, frame2)
		if isActive {
			onActive(i, score)
		}
	}
}

func diffFrames(frame1, frame2, diff []byte) {
	for i := 0; i < len(frame1); i++ {
		diff[i] = abs(frame1[i], frame2[i])
	}
}

type zone struct {
	// Index of masked pixels.
	// Even indexes are n masked pixels in row.
	// Odd indexes are n unmasked pixels in a row.
	maskIndex []int

	zoneSize  int
	frameSize int

	sensitivity  uint8
	thresholdMin float64
	thresholdMax float64
}

func newZone(width int, height int, config zoneConfig) *zone {
	polygon := convertPolygon(width, height, config.Area)
	maskImage := ffmpeg.CreateInvertedMask(width, height, polygon)
	mask, zoneSize := parseMaskImage(maskImage)

	var check bool
	var index []int

	if !mask[0] {
		check = true
		index = []int{0}
	}

	for _, maskPixel := range mask {
		if check == maskPixel {
			// Last item++.
			index[len(index)-1]++
		} else {
			index = append(index, 1)
			check = maskPixel
		}
	}

	return &zone{
		maskIndex:    index,
		zoneSize:     zoneSize,
		frameSize:    width * height,
		sensitivity:  uint8(math.Round(config.Sensitivity * 2.56)),
		thresholdMin: config.ThresholdMin,
		thresholdMax: config.ThresholdMax,
	}
}

func (z zone) checkDiff(diff []uint8) (float64, bool) {
	var nChangedPixels int
	var pos int
	var check bool
	for _, index := range z.maskIndex {
		if check {
			for i := 0; i < index; i++ {
				if diff[pos] >= z.sensitivity {
					nChangedPixels++
				}
				pos++
			}
		} else {
			pos += index
		}
		check = !check
	}

	percentChanged := (float64(nChangedPixels) / float64(z.zoneSize)) * 100
	isActive := percentChanged > z.thresholdMin && percentChanged < z.thresholdMax

	return percentChanged, isActive
}

func abs(x, y uint8) uint8 {
	if x < y {
		return y - x
	}
	return x - y
}

// convertPolygon percent to absolute values.
func convertPolygon(w int, h int, area area) ffmpeg.Polygon {
	polygon := make(ffmpeg.Polygon, len(area))
	for i, point := range area {
		px := point[0]
		py := point[1]
		polygon[i] = [2]int{
			int(float64(w) * (float64(px) / 100)),
			int(float64(h) * (float64(py) / 100)),
		}
	}
	return polygon
}

func parseMaskImage(img image.Image) ([]bool, int) {
	var mask []bool
	var zoneSize int

	max := img.Bounds().Max
	for y := 0; y < max.Y; y++ {
		for x := 0; x < max.X; x++ {
			_, _, _, a := img.At(x, y).RGBA()
			if a == 0 {
				zoneSize++
				mask = append(mask, false)
			} else {
				mask = append(mask, true)
			}
		}
	}
	return mask, zoneSize
}

File: plugins/motion/zone.rs
// SPDX-License-Identifier: GPL-2.0-or-later

use crate::config::ZoneConfig;
use common::recording::{create_inverted_mask, denormalize_polygon};

#[derive(Debug, PartialEq)]
pub struct Zones(pub Vec<Zone>);

impl Zones {
    pub fn analyze(&self, frame1: &[u8], frame2: &[u8], diff: &mut [u8]) -> Detections {
        diff_frames(frame1, frame2, diff);

        let mut detections = Vec::new();
        for (i, zone) in self.0.iter().enumerate() {
            let (score, is_active) = zone.check_diff(diff);
            if is_active {
                detections.push((i, score));
            }
        }
        detections
    }
}

pub type Detections = Vec<(usize, f32)>;

fn diff_frames(frame1: &[u8], frame2: &[u8], diff: &mut [u8]) {
    for i in 0..frame1.len() {
        diff[i] = u8::abs_diff(frame1[i], frame2[i]);
    }
}

#[allow(clippy::struct_field_names)]
#[derive(Debug, PartialEq)]
pub struct Zone {
    mask: Vec<bool>,

    zone_size: u64,
    frame_size: i64,

    sensitivity: u8,
    threshold_min: f32,
    threshold_max: f32,
}

impl Zone {
    #[allow(
        clippy::cast_sign_loss,
        clippy::cast_possible_truncation,
        clippy::as_conversions
    )]
    pub(crate) fn new(width: u16, height: u16, config: &ZoneConfig) -> Self {
        let polygon = denormalize_polygon(&config.area, width, height);
        let mask_image = create_inverted_mask(&polygon, width, height);
        let (mask, zone_size) = parse_mask_image(&mask_image);

        Zone {
            mask,
            zone_size,
            frame_size: i64::from(width) * i64::from(height),
            sensitivity: (config.sensitivity * 2.56) as u8,
            threshold_min: config.threshold_min,
            threshold_max: config.threshold_max,
        }
    }

    #[allow(clippy::cast_precision_loss, clippy::as_conversions)]
    fn check_diff(&self, diff: &[u8]) -> (f32, bool) {
        let mut n_changed_pixels = 0;
        for (i, is_masked) in self.mask.iter().enumerate() {
            if *is_masked && diff[i] >= self.sensitivity {
                n_changed_pixels += 1;
            }
        }

        let percent_changed = (n_changed_pixels as f32 / self.zone_size as f32) * 100.0;

        let is_active =
            percent_changed > self.threshold_min && percent_changed < self.threshold_max;

        (percent_changed, is_active)
    }
}

fn parse_mask_image(img: &Vec<Vec<bool>>) -> (Vec<bool>, u64) {
    let mut mask = Vec::new();
    let mut zone_size = 0;
    for row in img {
        for pixel in row {
            mask.push(*pixel);
            if *pixel {
                zone_size += 1;
            }
        }
    }
    (mask, zone_size)
}

#[allow(clippy::float_cmp, clippy::needless_pass_by_value)]
#[cfg(test)]
mod tests {
    use super::*;
    use common::{recording::normalize, PointNormalized};
    use pretty_assertions::assert_eq;
    use test_case::test_case;

    struct Case<'a> {
        name: &'a str,
        config: ZoneConfig,
        frame1: &'a [u8],
        frame2: &'a [u8],
        want: f32,
        is_active: bool,
    }

    fn p(x: u16, y: u16) -> PointNormalized {
        PointNormalized {
            x: normalize(x, 100),
            y: normalize(y, 100),
        }
    }

    #[test_case(
        "100%",
        ZoneConfig {
            enable: false,
            sensitivity: 8.0,
            threshold_min: 0.0,
            threshold_max: 90.0,
            area: vec![p(0, 0), p(100, 0), p(100, 100), p(0, 100)],
        },
        &[0, 0, 0, 0],
        &[255, 255, 255, 255],
        100.0,
        false
    )]
    #[test_case(
        "50%",
        ZoneConfig {
            enable: false,
            sensitivity: 8.0,
            threshold_min: 49.9,
            threshold_max: 50.1,
            area: vec![p(0, 0), p(100, 0), p(100, 100), p(0, 100)],
        },
        &[0, 0, 0, 0],
        &[255, 255, 0, 0],
        50.0,
        true
    )]
    #[test_case(
        "0%",
        ZoneConfig {
            enable: false,
            sensitivity: 8.0,
            threshold_min: 10.0,
            threshold_max: 0.0,
            area: vec![p(0, 0), p(100, 0), p(100, 100), p(0, 100)],
        },
        &[0, 0, 0, 0],
        &[0, 0, 0, 0],
        0.0,
        false
    )]
    fn test_compare_frames2(
        name: &str,
        config: ZoneConfig,
        frame1: &[u8],
        frame2: &[u8],
        want: f32,
        is_active: bool,
    ) {
        let zone = Zone::new(2, 2, &config);
        let mut diff = [0; 4];
        diff_frames(frame1, frame2, &mut diff);

        let (got, is_active2) = zone.check_diff(&diff);
        assert_eq!(want, got, "{0}", name);
        assert_eq!(is_active, is_active2, "{0}", name);
    }

    #[test]
    fn test_compare_frames() {
        let cases: Vec<Case> = vec![
            Case {
                name: "sensitivity",
                config: ZoneConfig {
                    enable: false,
                    sensitivity: 50.0,
                    threshold_min: 100.0,
                    threshold_max: 0.0,
                    area: vec![p(0, 0), p(100, 0), p(100, 100), p(0, 100)],
                },
                frame1: &[0, 0, 0, 0],
                frame2: &[127, 127, 127, 128],
                want: 25.0,
                is_active: false,
            },
            Case {
                name: "area 50%",
                config: ZoneConfig {
                    enable: false,
                    sensitivity: 8.0,
                    threshold_min: 100.0,
                    threshold_max: 0.0,
                    area: vec![p(0, 0), p(50, 0), p(50, 100), p(0, 100)],
                },
                frame1: &[0, 0, 0, 0],
                frame2: &[255, 0, 255, 0],
                want: 100.0,
                is_active: false,
            },
            Case {
                name: "area 25%",
                config: ZoneConfig {
                    enable: false,
                    sensitivity: 8.0,
                    threshold_min: 100.0,
                    threshold_max: 0.0,
                    area: vec![p(50, 0), p(100, 0), p(100, 50), p(50, 50)],
                },
                frame1: &[0, 0, 0, 0],
                frame2: &[0, 255, 0, 0],
                want: 100.0,
                is_active: false,
            },
        ];

        for tc in cases {
            let zone = Zone::new(2, 2, &tc.config);
            let mut diff = [0; 4];
            diff_frames(tc.frame1, tc.frame2, &mut diff);

            let (got, is_active) = zone.check_diff(&diff);
            assert_eq!(tc.want, got, "{0}", tc.name);
            assert_eq!(tc.is_active, is_active, "{0}", tc.name);
        }
    }

    /*func BenchmarkDetector(b *testing.B) {
        width := 500
        height := 500
        frameSize := width * height
        frame1 := bytes.Repeat([]byte{0}, frameSize)
        frame2 := bytes.Repeat([]byte{255}, frameSize)
        diff := make([]byte, frameSize)

        newTestZone := func(area area) *zone {
            return newZone(
                width,
                height,
                zoneConfig{
                    Enable:       true,
                    Sensitivity:  8,
                    ThresholdMin: 10,
                    ThresholdMax: 100,
                    Area:         area,
                },
            )
        }

        zones := zones{
            // Full frame.
            newTestZone(area{{0, 0}, {100, 0}, {100, 100}, {0, 100}}),
            // Large diamond 50%.
            newTestZone(area{{50, 0}, {100, 50}, {50, 100}, {0, 50}}),
            // Medium diamond.
            newTestZone(area{{50, 25}, {75, 50}, {50, 75}, {25, 50}}),
        }

        var zone int
        var score float64
        var active bool
        b.ResetTimer()
        for i := 0; i < b.N; i++ {
            onActive := func(zone int, s float64) {
                score = s
            }
            zones.analyze(frame1, frame2, diff, onActive)
        }
        _, _, _ = zone, score, active

    }*/
}
